### PR TITLE
Add GCC 8 support

### PR DESCRIPTION
**Pull request by xuyiqun-learner**
After compiled with gcc8.0, I found it can't work normally because c++ set filesystem as single module under gcc8.0

### CODE REVIEW

**Review by balamurugana — CHANGES_REQUESTED** [review]
1. What errors are you getting?
2. What is your gcc version?
3. What is your OS and architecture?

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -18,6 +18,12 @@ list(APPEND SRCS args.cc baseclient.cc client.cc http.cc request.cc response.cc
 add_library(miniocpp STATIC ${SRCS})
 target_link_libraries(miniocpp ${requiredlibs})
 
+if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU"
+    AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 8.0
+    AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9.0 AND NOT MINGW)
+    target_compile_options(miniocpp PUBLIC -lstdc++fs)
```
Comment (harshavardhana) [inline src/CMakeLists.txt:24]: This should be 

```
target_link_libraries(miniocpp stdc++fs)
```

**Comment by xuyiqun-learner** [comment]
> 1. What errors are you getting?
> 2. What is your gcc version?
> 3. What is your OS and architecture?

1.invok miniocpp lib but couldn't work 
2. gcc-8.4.0
3. Ubuntu18.04 , aarch64

File: src/CMakeLists.txt (hunk 1 above)
Comment (xuyiqun-learner) [reply src/CMakeLists.txt:24]: Yep, just thought target_ compile_options had same effect

**Comment by balamurugana** [comment]
> > 1. What errors are you getting?
> > 2. What is your gcc version?
> > 3. What is your OS and architecture?
> 
> 1.invok miniocpp lib but couldn't work 2. gcc-8.4.0 3. Ubuntu18.04 , aarch64

Please paste exact error messages.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -18,6 +18,12 @@ list(APPEND SRCS args.cc baseclient.cc client.cc http.cc request.cc response.cc
 add_library(miniocpp STATIC ${SRCS})
 target_link_libraries(miniocpp ${requiredlibs})
 
+if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU"
```
Comment (balamurugana) [inline src/CMakeLists.txt:21]: The actual fix is
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 73eee94..5389961 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -40,13 +40,16 @@ IF (${CMAKE_BINARY_DIR} STREQUAL ${CMAKE_SOURCE_DIR})
   MESSAGE(FATAL_ERROR "Please create a separate build directory and call CMake again")
 ENDIF (${CMAKE_BINARY_DIR} STREQUAL ${CMAKE_SOURCE_DIR})
 
-IF(CMAKE_COMPILER_IS_GNUCC )
-  set(CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} -Werror")
-ENDIF()
-
 # Look for required libraries
 SET(requiredlibs)
 
+IF(CMAKE_COMPILER_IS_GNUCC)
+  set(CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} -Werror")
+  IF(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 8.0 AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9.0 AND NOT MINGW)
+    list(APPEND requiredlibs stdc++fs)
+  ENDIF()
+ENDIF()
+
 find_package(CURL REQUIRED)
 IF(CURL_FOUND)
   INCLUDE_DIRECTORIES(${CURL_INCLUDE_DIRS})
```

**Review by balamurugana — CHANGES_REQUESTED** [review]

**Comment by xuyiqun-learner** [comment]
> > > 1. What errors are you getting?
> > > 2. What is your gcc version?
> > > 3. What is your OS and architecture?
> > 
> > 
> > 1.invok miniocpp lib but couldn't work 2. gcc-8.4.0 3. Ubuntu18.04 , aarch64
> 
> Please paste exact error messages.

It was strange that program didn't crash or report error msg, just invoked miniocpp library not response

**Review by balamurugana — CHANGES_REQUESTED** [review]
You would need to apply the diff correctly.

File: src/CMakeLists.txt (hunk 2 above)
Comment (xuyiqun-learner) [reply src/CMakeLists.txt]: > The actual fix is
> 
> ```diff
> diff --git a/CMakeLists.txt b/CMakeLists.txt
> index 73eee94..5389961 100644
> --- a/CMakeLists.txt
> +++ b/CMakeLists.txt
> @@ -40,13 +40,16 @@ IF (${CMAKE_BINARY_DIR} STREQUAL ${CMAKE_SOURCE_DIR})
>    MESSAGE(FATAL_ERROR "Please create a separate build directory and call CMake again")
>  ENDIF (${CMAKE_BINARY_DIR} STREQUAL ${CMAKE_SOURCE_DIR})
>  
> -IF(CMAKE_COMPILER_IS_GNUCC )
> -  set(CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} -Werror")
> -ENDIF()
> -
>  # Look for required libraries
>  SET(requiredlibs)
>  
> +IF(CMAKE_COMPILER_IS_GNUCC)
> +  set(CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} -Werror")
> +  IF(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 8.0 AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9.0 AND NOT MINGW)
> +    list(APPEND requiredlibs stdc++fs)
> +  ENDIF()
> +ENDIF()
> +
>  find_package(CURL REQUIRED)
>  IF(CURL_FOUND)
>    INCLUDE_DIRECTORIES(${CURL_INCLUDE_DIRS})
> ```

I had modfied as this.

Comment (balamurugana) [reply src/CMakeLists.txt]: The modification is wrong. There is no need to have two separate `IF(CMAKE_COMPILER_IS_GNUCC)`

**Comment by xuyiqun-learner** [comment]
> You would need to apply the diff correctly.

got it.

**Review by balamurugana — APPROVED** [review]

**Pull request merged**